tests/lib/rules/require-v-if-else-key.js

```diff
@@ -0,0 +1,159 @@
+
+const tester = new RuleTester({
+  parser: require.resolve('vue-eslint-parser'),
+  parserOptions: {
+    ecmaVersion: 2020,
+    sourceType: 'module'
+  }
+})
+
+tester.run('require-v-if-else-key', rule, {
+  valid: [
+    {
+      filename: 'test.vue',
+      code: `
+      <template>
+        <div>
+          <div v-if="conditional" :key="foo">foo</div>
+          <div v-else :key="bar">bar</div>
+        </div>
+      </template>
+      `
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <template>
+        <div>
+          <div v-if="conditional" key="foo">foo</div>
+          <div v-else key="bar">bar</div>
+        </div>
+      </template>
+      `
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <template>
+        <div>
+          <div v-if="conditional">foo</div>
+        </div>
+      </template>
+      `
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <template>
+        <div>
+          <div v-if="conditional">foo</div>
+          <span v-else>bar</span>
+        </div>
+      </template>
+      `
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <template>
+        <div>
+          <div v-if="conditional">foo</div>
+          <span v-else-if="conditional">bar</span>
+        </div>
+      </template>
+      `
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <template>
+        <div>
+          <div v-if="conditional" key="foo">foo</div>
+          <div v-else-if="conditional" key="bar">bar</span>
+        </div>
+      </template>
+      `
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <template>
+        <div>
+          <div v-if="conditional" key="one">foo</div>
+          <div v-else-if="conditional" key="two">bar</span>
+          <div v-else key="three">else</div>
+          <div v-if="otherConditional" key="four">baz</div>
+          <div v-else-if="otherConditional" key="five">quux</span>
+          <div v-else key="six">otherElse</div>
+        </div>
+      </template>
+      `
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <template>
+        <div>
+          <template v-if="conditional">
+            <div>foo</div>
+            <div>bar</span>
+          </template>
+        </div>
+      </template>
+      `
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <template>
+        <div>
+          <div v-if="conditional">foo</div>
+          <template v-else>
+            <div>bar</div>
+            <div>baz</span>
+          </template>
+        </div>
+      </template>
+      `
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <template>
+        <div>
+          <template v-if="conditional">
+            <div>foo</div>
+            <div>bar</span>
+          </template>
+          <template v-else>
+            <div>baz</div>
+            <div>quux</span>
+          </template>
+        </div>
+      </template>
+      `
```

Comment:
Why this is valid? I think this should be reported.

Reply:
I must have misinterpreted [your comment](https://github.com/vuejs/eslint-plugin-vue/pull/1845#discussion_r850099546):

> Note that vue2 does not allow you to place a key in <template>.

If no keys are allowed, mustn't we allow templates without keys?

Reply:
In Vue2 need to put the key on the actual element, not the `<template>`.

e.g.

```vue
<template v-if="foo">
  <div key="a" />
</template>
<template v-else>
  <div key="b" />
</template>
```

```vue
<template v-if="foo">
  <div key="a" />
  <div key="b" />
</template>
<template v-else>
  <div key="c" />
  <div key="d" />
</template>
```

However, with Vue3 you can do the following:



```vue
<template v-if="foo" :key="a">
  <div />
</template>
<template v-else :key="b">
  <div />
</template>
```

```vue
<template v-if="foo" :key="a">
  <div />
  <div />
</template>
<template v-else :key="b">
  <div />
  <div />
</template>
```